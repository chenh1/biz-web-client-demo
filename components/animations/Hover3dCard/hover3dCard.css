.card-3d {
  transition: transform 0.1s ease;
  transform-style: preserve-3d;
  will-change: transform;
}

.card-3d > * {
  position: relative;
  transition: transform 0.3s ease;
}

.card-3d:hover > * {
  z-index: 1;
  transform: translateZ(12px);
}

.card-3d::before {
  content: "";
  position: absolute;
  height: 100%;
  width: 100%;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
}

@media (prefers-reduced-motion) {
  .card-3d {
    transform: none !important;
  }
}