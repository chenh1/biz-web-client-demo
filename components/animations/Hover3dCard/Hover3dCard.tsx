'use client'

import React, { useRef } from 'react';
import './hover3dCard.css';

const THRESHOLD = 3;

export const Hover3dCard: React.FC<{ children?: string | JSX.Element | JSX.Element[] }> = ({ children }) => {
  const containerRef = useRef<HTMLDivElement>(null);

  const handleHover = (e) => {
    const { clientX, clientY, currentTarget } = e;
    const { clientWidth, clientHeight, offsetLeft, offsetTop } = currentTarget;
  
    const horizontal = (clientX - offsetLeft) / clientWidth;
    const vertical = (clientY - offsetTop) / clientHeight;

    const rotateX = (THRESHOLD / 2 - horizontal * THRESHOLD).toFixed(2);
    const rotateY = (vertical * THRESHOLD - THRESHOLD / 2).toFixed(2);

    if (!!containerRef.current) {
      containerRef.current.style.transform = `perspective(${clientWidth}px) rotateX(${rotateY}deg) rotateY(${rotateX}deg) scale3d(1, 1, 1)`;
    }
  }
  
  const resetStyles = (e) => {
    if (!!containerRef.current) {
      containerRef.current.style.transform = `perspective(${e.currentTarget.clientWidth}px) rotateX(0deg) rotateY(0deg)`;
    }
  }

  return (
    <div data-testid="hover-3d-card" ref={containerRef} onMouseMove={handleHover} onMouseLeave={resetStyles} className="card-3d">
      {children}
    </div>
  )
}