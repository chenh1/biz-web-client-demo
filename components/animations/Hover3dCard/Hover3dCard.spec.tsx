import { Hover3dCard } from "./Hover3dCard";
import { render, waitFor, fireEvent } from '@testing-library/react';
import { JSDOM } from "jsdom"
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

it('creates a component with a 3d card effect on hover', async () => {
  const { getByTestId, asFragment } = render(
    <Hover3dCard>
      <p>Hover3dCard</p>
    </Hover3dCard>
  );

  await waitFor(() => {
    expect(asFragment()).toMatchSnapshot();
  });
});


it('subtley rotates the card with a 3d effect on mouse over', async () => {
  const { getByTestId, asFragment } = render(
    <Hover3dCard>
      <p>Hover3dCard</p>
    </Hover3dCard>
  );

  const el = getByTestId('hover-3d-card');
  fireEvent.mouseMove(el, { clientX: 50, clientY: 50 });

  await waitFor(() => {
    expect(asFragment()).toMatchSnapshot();
  });
});