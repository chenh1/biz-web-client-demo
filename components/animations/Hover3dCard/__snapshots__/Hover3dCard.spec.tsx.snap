// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a component with a 3d card effect on hover 1`] = `
<DocumentFragment>
  <div
    class="card-3d"
    data-testid="hover-3d-card"
  >
    <p>
      Hover3dCard
    </p>
  </div>
</DocumentFragment>
`;

exports[`subtley rotates the card with a 3d effect on mouse over 1`] = `
<DocumentFragment>
  <div
    class="card-3d"
    data-testid="hover-3d-card"
    style="transform: perspective(0px) rotateX(Infinitydeg) rotateY(-Infinitydeg) scale3d(1, 1, 1);"
  >
    <p>
      Hover3dCard
    </p>
  </div>
</DocumentFragment>
`;
